feat: Support for User Parts Lists

lukadd16 opened the pull request:
This PR addresses issue #7, fixes some bugs with the asynchronous functions and makes some very minor fixes to code formatting.

---

First and foremost is adding the support for parts lists that exist under the `https://pcpartpicker.com/user/username/saved/` URL. Behind the scenes, these parts lists are effectively the same as the ones that the API currently supports (`https://pcpartpicker.com/list/`), so it was merely a matter of updating the logic that determined the validity of a URL.

The RegEx used in the `get_list_links()` function (pypartpicker/regex.py) has been modified to match `pcpartpicker.com/list/code` OR `pcpartpicker.com/user/username/code`. Since usernames are of variable length, I used the following non-capture group `(?:[\\w]+)` which matches a string that contains at least one alphanumeric (including underscore) character (the `+` quantifier is what enables this variable length behaviour).

I also cleaned up the RegEx that checked for the country subdomain in both the `get_list_links()` and `get_product_links()` functions. So rather than having a conditional that checks for `country.pcpartpicker` OR `pcpartpicker`, instead I've opted to check for only `country.pcpartpicker` but with `country.` being defined as an optional field (using `?`).

---

Next was actually updating pypartpicker/Scraper.py so that it would accept `/user/` URLs. I noticed that in the following lines you opted *not* to use any form of RegEx even though you already had two functions that could check for the same thing.

https://github.com/QuaKe8782/pypartpicker/blob/2fadced708ef1b6201a7cd4dbc7747b950c7e5c1/pypartpicker/scraper.py#L86

https://github.com/QuaKe8782/pypartpicker/blob/2fadced708ef1b6201a7cd4dbc7747b950c7e5c1/pypartpicker/scraper.py#L236

My proposed implementation adds two new [private methods](https://www.geeksforgeeks.org/private-methods-in-python/): `__check_list_url()` and `__check_product_url()` which utilize the updated RegEx from earlier. If the URL string that was passed to the `fetch_list()` or `fetch_product()` methods is of the desired format, then the corresponding private method will return **True** (and your if statement will be skipped over). It doesn't really get much simpler/elegant than that in my opinion.

---

Some other notable changes worth mentioning:
* The three asynchronous methods all had a **serious bug that prevented them from executing**. A `TypeError` was being thrown due to not passing `self` as a parameter in the method signature (when within the method you were calling methods from the class). I don't mean to sound arrogant, but this is something that you would have easily caught had you actually [tested](https://medium.com/@AgariInc/strategies-for-testing-async-code-in-python-c52163f2deab) the methods before committing.
* The `make_soup()` method has been converted to a private method (under the assumption that it is just a helper method that a user shouldn't be able to access/call on their own outside of the Scraper class).
* Updated all calls to the `make_soup()` method to account for the above change.
* Raise a [ValueError](https://docs.python.org/3/library/exceptions.html#ValueError) instead of a bare Exception when `fetch_list()` detects an invalid PCPP list URL (since this has to do with invalid user input).
* Fixed line spacing between classes & methods
* Converted use of `!=` in conditionals to more proper `is not`

lukadd16 commented:
Also, if you noticed I used the `feat:` scope in the PR title (which corresponds to the Conventional Commits standard that I mentioned in an issue before). And in terms of Semantic Versioning these changes would constitute a bump to the MINOR version (thus `1.8`).

lukadd16 commented:
I can also assure you that I have sufficiently tested each of the new and modified methods to ensure that they work as intended.

QuaKe8782 commented:
Looks good. Not sure why the async methods worked for me when I tested them. Thanks for all your help.

lukadd16 commented:
No problem, glad I could be of service.

Just one more thing, it's good that you're incorporating SemVer now, but the notes you attach to the new releases aren't very descriptive. E.g. Reading "Main changes here are features from the lukadd16 pull request." forces me to have to search for this repository, open up the PR tab and then hopefully find the PR in question. Now that's not to say you should write a whole paragraph summarizing these changes, but even just listing something like

## PyPartpicker v.1.8.0
### 🚀 Features:
* Added support for User Parts Lists (#8)

### :bug: Bug Fixes:
* Fixed TypeError bug with asynchronous methods (#8)

### 🤖 Chores
* New Versioning Scheme (all future releases will conform to the [Semantic Versioning](https://semver.org/) standard)

goes a long ways in describing to a programmer using your package what has changed + it gives them the option to look further into the corresponding PR/commit.

QuaKe8782 commented:
Alright I'll keep that in mind.
